A76-ring-hash-improvements.md

```diff
@@ -0,0 +1,236 @@
+hash for the request. If the use of this random hash triggers a connection
+attempt (according to the rules described in [A42: Picker
+Behavior][A42-picker-behavior] and updated in [A61: Ring Hash][A61-ring-hash]),
+then before queuing the pick, the picker will scan forward searching for a
+subchannel in `READY` state. If it finds a subchannel in `READY` state, the
+picker returns it.
+
+The following pseudo code describes the updated picker logic:
+
+```
+// Determine request hash.
+using_random_hash = false;
+if (config.request_hash_header.empty()) {
+  request_hash = call_attributes.hash;
+} else {
+  header = headers.find(config.request_hash_header);
+  if (header == null) {
+    using_random_hash = true;
+    request_hash = ComputeRandomHash();
+  } else {
+    request_hash = ComputeHash(header);
+  }
+}
+
+// Do pick based on hash.
+first_index = ring.FindIndexForHash(request_hash);
+requested_connection = false;
+for (i = 0; i < ring.size(); ++i) {
+  index = (first_index + i) % ring.size();
+  if (ring[index].state == READY) {
+    return ring[index].picker->Pick(...);
+  }
+  if (requested_connection) continue;
+  if (ring[index].state == IDLE) {
+    ring[index].endpoint.TriggerConnectionAttemptInControlPlane();
+    if (using_random_hash) {
+      requested_connection = true;
+      continue;
+    }
+    return PICK_QUEUE;
+  }
+  if (ring[index].state == CONNECTING) {
+    return PICK_QUEUE;
+  }
+}
+if (requested_connection) return PICK_QUEUE;
+return PICK_FAIL;
+```
+
+This behavior ensures that a single request creates at most one new connection,
+and that a request missing the header does not add extra latency in the common
+case where there is already at least one subchannel in `READY` state. It
+converges to picking a random host, since each request may create a new
+connection to a random endpoint.
+
+### Explicitly setting the endpoint hash key
+
+The `ring_hash` policy will be changed such that the hash key used for
+determining the locations of each endpoint on the ring will be extracted from a
+pre-defined endpoint attribute called `hash_key`. If this attribute is set, then
+the endpoint is placed on the ring by hashing its value. If this attribute is
+not set or empty, then the endpoint's first address is hashed, matching the
+current behavior. The locations of an existing endpoint on the ring is updated
+if its `hash_key` endpoint attribute changes.
+
+The xDS resolver, described in [A74][A74], will be changed to set the `hash_key`
+endpoint attribute to the value of [LbEndpoint.Metadata][LbEndpoint.Metadata]
+`envoy.lb` `hash_key` field, as described in [Envoy's documentation for the ring
+hash load balancer][envoy-ring-hash].
+
+### Temporary environment variable protection
+
+Explicitly setting the request hash key cannot possibly lead to problem with
+existing deployment because the new behavior requires setting a load balancing
+policy configuration field that did not exist before. Therefore, it is not gated
```

Comment:
That's not the condition. It's "whether remote I/O can trigger it" and it can. The reason to have an environment variable here is to handle bugs in implementations while they are being implemented. If there's a crasher bug, for example, you don't want to _prevent_ using the new feature because you might trigger old clients to crash.

Reply:
OK, so I guess for this one it also makes sense to have it be opt-in to start with. I added an environment variable condition called `GRPC_EXPERIMENTAL_RING_HASH_SET_REQUEST_HASH_KEY`. That would let users disable the feature should it cause problems if the source of the service config is causing an undesirable behavior. I am not familiar with the process of when we decided to remove this kind of feature gating environment variable. Please let me know if that seems right.

Reply:
The normal process is, "when it is adequately tested, enable it by default." C++ core tends to remove the env var at that point. In Java/Go we tend to wait a release with the env var there so it can be disabled if something goes awry. When the env var is removed doesn't matter; we care about when the default changes here.

But the main point is "when tested." In most other gRFC's that's our interop test framework; here, I guess it'll just be unit tests? For Go in practice it might be "when someone tests it with a service" where "someone" is "Antoine and friends".

Again, the main thing we want to avoid is a bug during the initial implementation that causes really bad client behavior such that you can't enable the feature in the future, lest you trigger broken clients.

Reply:
OK, that makes sense, and it also applies to the other environment variable then.

I'm happy to remove the environment variables when we have usage (we have internal usage of a fork of the existing ring hash that we'll be able to replace when this is implemented in Java and Go).